brew test, install, test-bot: add --keep-tmp option to retain staged files

apjanke opened the pull request:
- [X] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/.github/CONTRIBUTING.md) document?
- [X] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [X] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [X] Have you successfully ran `brew tests` with your changes locally?

This adds a `--keep-tmp` option to `brew test`, `brew install`, and `brew update-test` which lets you preserve the files for inspection after a failed test or install. It applies to the test dir for `brew test` and the build dir (with staged distribution) for `brew install`. Also adds it to `brew test-bot`, where it is propagated to the main `brew install` and `brew test` runs that test-bot does.

When `--interactive` or `--debug` are supplied, the temp file retention is done automatically if there is a build or test failure, even if `--keep-tmp` was not provided.

Enables the sandbox for interactive and debug use, now that we have a way of retaining those files for subsequent inspection. (The main reason the sandbox was disabled here is that it would prevent you from copying the staged temp files out to a new location, right? Though it seems like one could have just copied it to another location under Homebrew's cache, which was whitelisted.)

mikemcquaid commented:
Seems reasonable :+1: Just in case it's not too hard to roll in: I suspect a lot of people would like something similar for `brew install`.

mikemcquaid commented on Library/Homebrew/formula.rb:
```diff
@@ -1318,18 +1318,25 @@ def verify_download_integrity(fn)
-      ENV["HOME"] = @testpath
-      setup_home @testpath
-      test
+    preserve_files = ARGV.include?("--keep-files")
```
Maybe make this a `ARGV` function?

mikemcquaid commented on Library/Homebrew/formula.rb:
```diff
@@ -1318,18 +1318,25 @@ def verify_download_integrity(fn)
+      end
+    ensure
+      if preserve_files
+        ohai "Keeping temporary files"
```
Maybe `Kept`?

UniqMartin commented on Library/Homebrew/extend/ARGV.rb:
```diff
@@ -120,6 +120,10 @@ def dry_run?
 
+  def keep_files?
+    flag? "--keep-files"
+  end
```
I think this raises the question what files we're talking about (though that's documented). I think `keep_temp?` and `--keep-temp` would be more self-descriptive.

xu-cheng replied:
Also, probably want to use `include?` here to avoid introduce `-k` switch.

xu-cheng replied:
Another thought is if we only use it in `test`, we probably want to avoid adding this method to `ARGV`. Just use `ARGV.include? "--keep-files"` directly in `test.rb`

UniqMartin replied:
Mike proposed moving this to `ARGV` (see now hidden code comments) and I agree because like Mike I'd like to see this also being used by `brew install`. But good point on using `include?` instead of `flag?`!

apjanke replied:
Yeah, the intention is for `install` to eventually make use of it, too; I just don't have an implementation for it yet.

Changed to `include?`.

UniqMartin commented on Library/Homebrew/formula.rb:
```diff
@@ -1318,18 +1318,25 @@ def verify_download_integrity(fn)
+    begin
+      old_home = ENV["HOME"]
+      build, self.build = self.build, Tab.for_formula(self)
+      mktemp("#{name}-", { :preserve => preserve_files }) do
```
`{ :preserve => preserve_files }` should be `:preserve => preserve_files` instead. Also, if you used `ARGV.keep_files?` directly in place of `preserve_files`, there would be no need for a separate `begin`-`ensure`-`end` block and the diff would be more useful. (Just a suggestion.)

apjanke replied:
Did both these changes; diff looks nicer now.

UniqMartin commented on Library/Homebrew/formula.rb:
```diff
@@ -1318,18 +1318,25 @@ def verify_download_integrity(fn)
+      if preserve_files
+        ohai "Kept temporary files"
+        puts "Temporary files are saved at #{@testpath}"
+      end
```
I think a more succinct `ohai "Retained temporary files in #{@testpath}" if ARGV.keep_files?` would be sufficient.

apjanke replied:
`ohai` gets truncated to the terminal width so it could lose the path information, and that's necessary for this to be useful. And I didn't want to do just `puts` because without the `ohai` header it tends to get lost in the build output

UniqMartin replied:
Then make that: `ohai "Retained temporary files:", @testpath if ARGV.keep_files?`. `ohai` accepts multiple inputs and all but the first are printed (without styling or truncation) via `puts`. (I find it annoying that `ohai` truncates its first argument, but that's not something that can be solved here.)

apjanke replied:
Cool; I didn't know about the multi-arg `ohai`. I kind of prefer the current output, though. And I added a `!@testpath.nil?` check, so the one-line form becomes pretty long.

```
ohai "Kept temporary files", @testpath if ARGV.keep_files? && !@testpath.nil?
```

UniqMartin commented:
:+1: On the general idea (modulo my code comments) and would love to see this for `brew install` (either here or in a separate PR).

xu-cheng commented:
> The --interactive option doesn't work with the sandbox,

Should we make `--interactive` not disabling sandbox if we have `--keep-file`?

apjanke commented:
Amended based on feedback.

Yeah, `--keep-files` does seem generic, since this is intended to apply specifically to the temporary working directory the test (or build, if `install` is added) is run in, and other files may be produced as a result of an `install` at least. `brew install --keep-files` definitely doesn't scan right.

Alternate name preferences? `--keep-temp-files`? `--keep-working-dir`? `--keep-staging`?

apjanke commented:
> > The --interactive option doesn't work with the sandbox,
>
> Should we make `--interactive` not disabling sandbox if we have `--keep-file`?

If that's the only reason it doesn't work now, yeah. Or even make `--interactive` cause the files to be kept when any exceptions happen, so they're preserved when needed for interactivity, but deleted if everything works fine.

UniqMartin commented:
> Alternate name preferences? `--keep-temp-files`? `--keep-working-dir`? `--keep-staging`?

I'm still a big fan of `--keep-temp` as suggested in my code comment. The working directories of both `install` and `test` are temporary (in my mental model), so that seems appropriate and succinct.

> Should we make `--interactive` not disabling sandbox if we have `--keep-file`?

Even if that was the case I'd find it useful to have access to the build directory of a formula after a regular, non-interactive `brew install` every once in a while.

mikemcquaid commented:
> I'm still a big fan of --keep-temp as suggested in my code comment.

Maybe even `--keep-tmp`

apjanke commented:
I figured out how to do `--keep-tmp` for `brew install`, too, and added it.

Changed the implementation so now you have a `mktemp_preserve` function you can call from within the `mktmpdir do` block, so you can decide dynamically in there whether to preserve it. And the output comes from `mktmpdir` itself, so you don't have to hold on to a `mktmpdir` results object at the point of the top `do` call and have redundant output code. This decoupling made doing it for `brew install` feasible. Other features can make use of this, too.

The implementation is not thread-safe; it'll give incorrect results if multiple threads are running `mktmpdir` at the same time. We don't have to worry about multithreading in `brew`, do we?

Made the `--keep-tmp` automatic when there is a build or test failure under `--interactive` or `--debug`. And since that works and gives us a convenient way of seeing in to sandboxed runs, this re-enables the sandbox for `--debug` and `--interactive` use. You can still also explicitly pass `--keep-tmp` to retain the temp files for any test or install run, regardless of its exit status.

Since `brew test-bot` runs combination of `install` and `test`, I added it to `test-bot` too, to propagate to its `install` and `test` runs. The limitation here is that since the retention is done by child `brew` processes whose output is not displayed, you never see the "retained temp file" messages if it happens; you just have to look at your temp directory afterwards.

Changed the prefix used by every `mktmpdir` to include a "-" for readability.

Switched name to `--keep-tmp`; that makes most sense to me too.

UniqMartin commented on Library/Homebrew/cmd/test-bot.rb:
```diff
@@ -584,7 +591,9 @@ def formula(formula_name)
-        test "brew", "test", "--verbose", formula_name if formula.test_defined?
+        shared_test_args = ["--verbose"]
+        shared_test_args << "--keep-tmp" if ARGV.keep_tmp?
+        test "brew", "test", *shared_test_args, formula_name if formula.test_defined?
```
Unfortunately, this doesn't work with Ruby 1.8. The array splat (`*`) only works when used with the last argument. (Same for other instances of this pattern further down.)

apjanke replied:
ARRRGH. Okay, will fix.

UniqMartin commented:
The changes are quite extensive and I haven't been able to review them completely, however:

> Changed the implementation so now you have a `mktemp_preserve` function you can call from within the `mktmpdir do` block, so you can decide dynamically in there whether to preserve it. And the output comes from `mktmpdir` itself, so you don't have to hold on to a `mktmpdir` results object at the point of the top `do` call and have redundant output code. This decoupling made doing it for `brew install` feasible. Other features can make use of this, too.
> 
> The implementation is not thread-safe; it'll give incorrect results if multiple threads are running `mktmpdir` at the same time. We don't have to worry about multithreading in `brew`, do we?

Couldn't those complications (and the use of class variables) be avoided by making `Mktemp#run` yield the `Mktemp` instance to the invoking code? It would allow writing something like this:

```ruby
mktemp("prefix") do |temp|
  temp.preserve if ARGV.keep_tmp?
  …
end
```

> Since `brew test-bot` runs combination of `install` and `test`, I added it to `test-bot` too, to propagate to its `install` and `test` runs. The limitation here is that since the retention is done by child `brew` processes whose output is not displayed, you never see the "retained temp file" messages if it happens; you just have to look at your temp directory afterwards.

Isn't this going to fill up `/tmp` for every formula that fails to build for some reason? I think this might be an issue for a continuously running `test-bot` instance because it would slowly run out of disk space. (Or am I misunderstanding this?)

UniqMartin commented on Library/Homebrew/extend/fileutils.rb:
```diff
@@ -6,42 +6,95 @@
+      @preserve = opts[:preserve]
+    end
 
+    def run()
```
It's customary to omit an empty argument list (even in a declaration like here).

apjanke commented:
Fixed the Ruby 1.8.7 compatibility.

> > I added it to test-bot too, to propagate to its install and test runs.
>
> Isn't this going to fill up /tmp for every formula that fails to build for some reason?

Only if you add `--keep-tmp` to the `test-bot` calls, which this PR does not. `test-bot` does not use `--interactive` or `--debug` in the `brew install` and `brew test` calls it does, so the automatic retention on failure is not activated. It's intended to be used for interactive `brew test-bot` runs where you're trying to locally reproduce and diagnose a failure.

> Couldn't those complications (and the use of class variables) be avoided by making Mktemp#run yield the Mktemp instance to the invoking code?

Yes, but it would be a bit complicated and have to also be done to some existing methods.

We would have to build that yielding signature in to `patch`, `Formula.brew`, `Resource.stage`, and maybe a couple other things, because not every point that wants to call `mktemp_preserve` is directly calling `mktemp`. For example, here:

```
    formula.brew do
      FileUtils.mktemp_preserve if ARGV.keep_tmp?
      formula.patch
```

There's a few intermediate calls between that `formula.brew` and what eventually passes the block along to `mktemp`. They'd all have to change their signatures, and I thought committing to that would be a bigger change in both implementation and internal API. The current approach also lets you structure things so you can do the `mktemp_preserve` call inside functions called from that block, or even in functions which may or may not be running inside a `mktemp` call.

Doing it with the block argument does fix the locality/multithreading issues, and it looks like more idiomatic Ruby. Maybe it's worth it. My sense says no, because the change would hit so many commonly used things in existing code. But maybe it is: passing the mktemp context to those blocks would be a more appropriate mechanism for associating the block with the temporary space it's running in.

Open to opinions here from people with more Ruby/`brew` experience.

apjanke commented:
Turns out the pass-to-yield design isn't as big of a change: there's a lot of `stage` methods in the codebase, but the ones on DownloadStrategy classes have a different signature that doesn't take blocks; they are just called from inside the Resource.stage method, which is the one we need to modify.

[Here's an implementation](https://github.com/apjanke/brew/commit/a2a453e6b51686a4e7291497e337795073ed52d7) that changes this PR's class methods to instance methods on a stage context object passed to the blocks that `mktemp` and `stage` yield to.

I think this yield-based control flow design works better. It requires threading another parameter through all the `stage do` and `mktemp do` and `brew do` calls. But it correctly localizes the staging area context to the call stack acting on it, avoiding globals and allowing the contexts at any level of the stack to be acted on as long as you have the right object reference.

xu-cheng commented on Library/Homebrew/extend/fileutils.rb:
```diff
@@ -6,42 +6,95 @@
+        # Reference from `man 2 open`
+        # > When a new file is created, it is given the group of the directory which
+        # contains it.
+        group_id = if HOMEBREW_BREW_FILE.grpowned?
```
The indention need to be aligned here. :wink:

xu-cheng commented:
Could you write test case for the new `mktemp` as well as `mktemp_preserve`?

Also would it be a better idea to call `mktemp_preserve` to just `preserve_tempfile`?

apjanke commented:
I've merged in an updated implementation using block arguments like Martin suggested. Eliminates the globals and reentrancy problems, and (at least to me) looks more like idiomatic Ruby. Left it as a separate commit so it's easy to see the differences; will squash before this eventually gets merged.

> Could you write test case for the new mktemp as well as mktemp_preserve?

Sure.

UniqMartin commented on Library/Homebrew/build.rb:
```diff
@@ -105,7 +105,8 @@ def install
     formula.extend(Debrew::Formula) if ARGV.debug?
 
-    formula.brew do
+    formula.brew do |f, stg|
```
Is `f` actually used? If not, prefix it with an underscore to make that more obvious (and silence RuboCop).

I think the `stg` is unnecessarily cryptic. What does it stand for? `stage`? What does it actually represent? To me that's a handle to the temporary directory, so something like `temp_directory` (or `staging_directory`, if you prefer) would be nice. (Obviously also applies to the rest of the code.)

apjanke replied:
Yeah, `stg` for "stage", as in "staging context"; the context that goes with `stage`-like methods. I didn't want to take up a whole argument slot with just the temporary directory; I think it should apply to staged temporary data besides just the temp directory. If `brew` or `stage` ended up creating other stuff, like database entries or receipt files or ephemeral logs in another directory, `preserve` could also apply to those. Don't think spelled-out `stage` can be used because a `stage` method might be called inside these.

`stage_context`, `stage_data`, `stage_files`? `tmp_data`, `tmp_context`, maybe?

UniqMartin replied:
I personally prefer `stage_context`, though `staging_context` sounds more correct to my non-native understanding of the English language.

> Don't think spelled-out `stage` can be used because a `stage` method might be called inside these.

A variable will always take precedence over a same-named method. There's no method calling going on in `|<stuff>|` because that's basically an argument list (like in method definitions).

apjanke replied:
Right, I meant, I was avoiding using that name for these `|<foo>|` block parameters, because I wanted to use the same parameter name for all of them, and in some of these blocks, a call to a `stage` method might actually occur, so we need to avoid the local variable masking the same-named method. And since `stage` is a method name used in this functional area, I thought it would be clearer to keep separate names rather than rely on the type precedence to sort things out.

> I personally prefer stage_context, though staging_context sounds more correct to my non-native understanding of the English language.

"Staging context" is how you'd say it in spoken conversation, but I think `stage_context` works better as a variable, too. If there were several things related to a `stage`, I'd prefix them all with `stage_`, instead of switching between `stage_`, `staged_` and `staging_`.

apjanke replied:
Fixed `_f`. Leaving it as `stg` for now; will think about your name suggestions.

UniqMartin replied:
About `stage`: I see. Thanks for clarifying! `stage_context` is fine with me. I'm happy with almost everything that is less cryptic than `stg`. :smile_cat:

UniqMartin commented on Library/Homebrew/extend/fileutils.rb:
```diff
@@ -6,42 +6,79 @@
+    include FileUtils
+
+    # Path to the tmpdir used in this run, as a Pathname.
+    attr_accessor :tmpdir
```
I think this attribute should be read-only, maybe even not exposed at all.

UniqMartin commented on Library/Homebrew/extend/fileutils.rb:
```diff
@@ -6,42 +6,79 @@
+    # Path to the tmpdir used in this run, as a Pathname.
+    attr_accessor :tmpdir
+    # Whether the files in tmpdir should be retained
+    attr_accessor :do_preserve
```
`@do_preserve` does not need to be publicly accessible. It is already exposed via `preserve` and `preserve?`.

UniqMartin commented on Library/Homebrew/extend/fileutils.rb:
```diff
@@ -6,42 +6,79 @@
+    attr_accessor :do_preserve
+
+    # Instructs this Mktemp to preserve the staged files
+    def preserve
```
Not entirely sure about Ruby conventions, but to me this feels like a method that deserves a bang, i.e. `preserve!`.

UniqMartin commented on Library/Homebrew/extend/fileutils.rb:
```diff
@@ -55,6 +92,7 @@ def mkdir(name, &_block)
       end
     end
   end
+
```
I'd prefer if the PR didn't introduce empty lines unrelated to the changed code.

UniqMartin commented on Library/Homebrew/extend/fileutils.rb:
```diff
@@ -67,6 +105,7 @@ def mkdir(name, &_block)
     # @private
     class Entry_
       alias_method :old_copy_metadata, :copy_metadata
+
```
I'd prefer if the PR didn't introduce empty lines unrelated to the changed code.

UniqMartin commented on Library/Homebrew/extend/fileutils.rb:
```diff
@@ -132,6 +171,7 @@ def xcodebuild(*args)
   ensure
     ENV.update(removed)
   end
+
```
I'd prefer if the PR didn't introduce empty lines unrelated to the changed code.

UniqMartin commented on Library/Homebrew/sandbox.rb:
```diff
@@ -8,18 +8,6 @@ def self.available?
-      opoo "The sandbox cannot be used in debug or interactive mode."
-      @printed_autodisable_warning = true
-    end
-  end
```
I haven't looked into this, but could there be other reasons why the sandbox cannot/couldn't be used in debug/interactive mode? I.e. what is the original reason for this limitation?

xu-cheng replied:
The reason it used to be disabled is that it's hard to retain file in interactive shell. So we are safely to remove them for now.

UniqMartin commented on Library/Homebrew/cmd/test.rb:
```diff
@@ -10,6 +10,9 @@
 #:    launched with access to IRB or a shell inside the temporary test directory.
 #:
+#:    If `--keep-tmp` is passed, the temporary files created for the test are
+#:    not deleted.
```
Is it an oversight that you didn't update the documentation for `install` or are you just postponing this until the PR is basically finalized?

apjanke replied:
Oversight; thanks.

mikemcquaid commented on Library/Homebrew/build.rb:
```diff
@@ -105,7 +105,8 @@ def install
     formula.extend(Debrew::Formula) if ARGV.debug?
 
-    formula.brew do
+    formula.brew do |_f, stg|
```
Not obvious what `_f` or `stg` are; can you use longer variables here?

mikemcquaid commented on Library/Homebrew/extend/fileutils.rb:
```diff
@@ -6,42 +6,77 @@
+  # or calling `do |stg| ... stg.preserve` in the block will suppress
+  # the deletion for the topmost mktemp on the call stack, if any.
+  def mktemp(prefix = name, opts = {})
+    Mktemp.new(prefix, opts).run do |stg|
```
`stg` to `staging`?

mikemcquaid replied:
or even `dir`?

mikemcquaid commented on Library/Homebrew/formula.rb:
```diff
@@ -1537,7 +1548,7 @@ def exec_cmd(cmd, args, out, logfn)
 
   def stage
-    active_spec.stage do
+    active_spec.stage do |rsrc, stg|
```
`resource, dir`

mikemcquaid commented:
A few feelings on readability and variable names but otherwise :+1:

xu-cheng commented on Library/Homebrew/cmd/unpack.rb:
```diff
@@ -39,7 +39,7 @@ def unpack
       f.brew do
         f.patch if ARGV.flag?("--patch")
-        cp_r getwd, stage_dir, :preserve => true
+        cp_r getwd, stage_dir, :preserve! => true
```
This seems to be a mistake. The preserve is a flag for cp_r.

xu-cheng commented on Library/Homebrew/extend/fileutils.rb:
```diff
@@ -6,42 +6,77 @@
+
+    def initialize(prefix = name, opts = {})
+      @prefix = prefix
+      @do_preserve = opts[:preserve!]
```
Personally, I rare see `!` used in option name.

UniqMartin replied:
I have to agree. I'm the one who suggested the bang (`!`) in a previous code comment, but in my mind I had restricted that to the `Mktemp#preserve!` method.

Also, the `do_` in `@do_preserve` seems unnecessary and carries no information. Let's make this just `@preserve`.

apjanke replied:
The `!` in the option name was a mistake; result of overzealous search-and-replace. Will fix.

apjanke commented:
Bringing this up again since Martin asked about it, and I don't know for sure: was inability to access the files the only reason the sandbox was disabled for `--debug` and `--interactive` use? I think that's so, and as such, this PR turns the sandbox back on for those cases. But want to make sure I'm not overlooking something else.

apjanke commented:
Amended again:
* Added a unit test and fixed related tests
* Addressed naming concerns by spelling everything out; picked `staging` for the long block parameter name
* Fixed erroneous `:preserve!` that xu-cheng caught
* Renamed "preserve" to "retain" everywhere in this feature

I think "retain" better reflects what it's actually doing. "Preserve" usually asks a function or command to keep things in the state they were before the command was run, which isn't quite what's going on here. It's closer to "document retention" than "document preservation". And the comments & output were already using "retain"; may as well be consistent.

I think this might be ready to go now.

UniqMartin commented on Library/Homebrew/extend/fileutils.rb:
```diff
@@ -6,42 +6,83 @@
-      Process.gid
+  # recursively delete the temporary directory. Passing opts[:retain]
+  # or calling `do |staging| ... staging.retain!` in the block will suppress
+  # the deletion for the topmost mktemp on the call stack, if any.
```
The piece about the topmost `mktemp` and the call stack no longer applies, right?

apjanke replied:
Right. Fixed.

UniqMartin commented on Library/Homebrew/test/test_patching.rb:
```diff
@@ -122,7 +122,9 @@ def test_single_patch_dsl_with_incorrect_strip
-        end.brew(&:patch)
+        end.brew do |f, _staging|
+          f.patch
+        end
```
For both brevity and clarity, I would make that `end.brew { |formula, _staging| formula.patch }`. (The shorter `f` seems also acceptable, given how widely that is used in our code base.)

apjanke replied:
This I don't know about... brew's codebase seems to really avoid the `{ ... }` syntax except for short one liners, and this doesn't seem to count as a one-liner since it's attached to that other block. I'll look around and see if I can find similar examples in the existing code.

apjanke replied:
Well, the GitHub Ruby Style Guide says to use `{ ... }` for single-line blocks, but to avoid multiline chaining like this entirely, and to refactor a multi-line chain in to methods. (It's in the bullet point that starts with "Prefer {...} over do...end for single-line blocks".)

Since the multi-line chain was there before, and I'm not sure how to refactor it away without (imho) hurting readability, I'm going to switch this over to `{ ... }` and leave it as a chain. More thorough style fixes can be done in another PR; there's a bunch of other items that `brew style` also complains about in this file.

UniqMartin replied:
Thanks! Yeah, we still have a lot of work to do to make RuboCop entirely happy with out code, even though we have already disabled or relaxed a lot of its rules in our configuration. :wink:

UniqMartin commented on Library/Homebrew/test/test_patching.rb:
```diff
@@ -136,7 +138,9 @@ def test_single_patch_dsl_with_incorrect_strip_with_apply
             apply APPLY_A
           end
-        end.brew(&:patch)
+        end.brew do |f, _staging|
```
For both brevity and clarity, I would make that `end.brew { |formula, _staging| formula.patch }`. (The shorter `f` seems also acceptable, given how widely that is used in our code base.)

UniqMartin commented on Library/Homebrew/test/test_patching.rb:
```diff
@@ -136,7 +138,9 @@ def test_single_patch_dsl_with_incorrect_strip_with_apply
-        end.brew(&:patch)
+        end.brew do |f, _staging|
+          f.patch
+        end
```
Same here.

UniqMartin commented on Library/Homebrew/test/test_patching.rb:
```diff
@@ -234,7 +238,9 @@ def test_single_patch_dsl_with_apply_enoent_fail
-        end.brew(&:patch)
+        end.brew do |f, _staging|
+          f.patch
+        end
```
Same here.

UniqMartin commented on Library/Homebrew/utils.rb:
```diff
@@ -173,7 +173,7 @@ def interactive_shell(f = nil)
   elsif $?.exited?
     puts "Aborting due to non-zero exit status"
-    exit $?.exitstatus
+    raise "Aborted due to non-zero exit status (#{$?.exitstatus})"
```
It looks like the `puts` directly above this line is now obsolete.

apjanke replied:
Yeah, it leaves redundant output. Removed.

UniqMartin commented:
Left some nitpicking code comments, but nothing really substantial. This means I really like how the PR has progressed and what the code looks like now. Also like the move from “preserve” to “retain”.

:+1: once the sandbox-related question is answered.

apjanke commented:
Thanks! I'm happy with it too. Glad you pushed for the block-argument form; I think it's a better design than the globals.

I addressed the nitpicks, and rebased on master to pick up the `man`-related changes.

xu-cheng confirmed in a line comment that the sandbox disabling was just file-related: "The reason it used to be disabled is that it's hard to retain file in interactive shell. So we can safely remove them now."

If no objections, I'll pull this in this evening.

apjanke commented:
Merged. Thanks for your feedback!

apjanke commented:
Note: this PR caused breakage found in #529, because in older versions of Ruby, a multi-argument call to a block which only has a single parameter causes all the arguments to be condensed into an array, instead of omitting the later arguments. This broke the interface for `stage` as used in some `Formula` definitions. We need to keep the interface for stage to using only a single argument in its block.

Working on a fix that goes back to the old signature.
